Variants by region

pvighi opened the pull request:
This PR adds the option of providing a list of country groups a variant applies to. When picking a variant on a specific request only variants that match the country group of the request will be considered. Variants defined with an empty country group filter will apply to all country groups.

pvighi committed:
Region specific variants

pvighi committed:
removed unused Variants

pvighi committed:
undo change that was introduced in a different pull request

pvighi commented on app/views/support/ABTest.scala:
```diff
@@ -90,27 +107,29 @@ case class ChosenVariants(v1: AmountHighlightTest.Variant, v2: MessageCopyTest.V
+
+    def pickRandomly: test.Variant = {
+      val n = scala.util.Random.nextDouble()
+      test.weightedVariantsFor(countryGroup).dropWhile(_._1 < n).head._2
```
This will crash horribly if we forget to include a variant that applies to the current countryGroup. 
If we want to avoid making this mistake we could either force the app to crash on startup if we end up with an empty list for any of the supported countryGroups or do headOption instead of head here and fall back to some default value

pvighi committed:
add check on startup to detect invalid test configurations

alexduf commented on app/views/support/ABTest.scala:
```diff
@@ -6,26 +6,51 @@ import java.net.URLEncoder
-  val weightedVariants = cdf.zip(variants.list)
+
+  def ValidateVariants = {
+    val countryWithNoVariants = variantsByCountry.collectFirst { case (country, variants) if variants.isEmpty => country.name }
```
elegant :100:

alexduf commented on app/views/support/ABTest.scala:
```diff
@@ -90,27 +115,33 @@ case class ChosenVariants(v1: AmountHighlightTest.Variant, v2: MessageCopyTest.V
+  def pickVariant[A](countryGroup: CountryGroup, request: Request[A], test: TestTrait): test.Variant = {
+
+    def pickRandomly: test.Variant = {
+      val n = scala.util.Random.nextDouble()
```
while you're at it, could you add an `import scala.util.Random` ?

alexduf commented:
Looks good :+1: you probably want another pair of :eyes: on that: https://github.com/guardian/contributions-frontend/pull/20/files#diff-114818ea49160145b7f0d015ed576610R21

pvighi committed:
merge and fix conflicts with latest master version

pvighi committed:
fixed merge conflicts after updating from master

pvighi committed:
changed country group filter to a set

pvighi committed:
minor formatting change

alexduf commented:
:+1:

rtyley commented on app/views/support/ABTest.scala:
```diff
@@ -6,26 +6,51 @@ import java.net.URLEncoder
+    errorMessage.foreach(m => throw new IllegalStateException(m))
+  }
+
+  val variantsByCountry: Map[CountryGroup, List[Variant]] = {
```
The `List` type can probably be `Set`?

rtyley commented on app/views/support/ABTest.scala:
```diff
@@ -6,26 +6,51 @@ import java.net.URLEncoder
+
+  val weightedVariantsByCountry = {
+
+    def addWeights(filteredVariants: List[Variant]) = {
```
I like all this stuff! But because it's not that simple, maybe it's worth adding a couple of unit tests for this? Asserting that the cumulative distribution function adds up to 1, that sort of thing?

pvighi replied:
Ok I'll add a unit test for this, and I will remove the runtime check with the ugly exception that I do in validateVariants and move it to this test as well

markjamesbutler commented on app/views/support/ABTest.scala:
```diff
@@ -6,26 +6,51 @@ import java.net.URLEncoder
+    val countryWithNoVariants = variantsByCountry.collectFirst { case (country, variants) if variants.isEmpty => country.name }
+    val errorMessage = countryWithNoVariants.map(country => s"${getClass.getName}: No variant defined for $country.")
+
+    errorMessage.foreach(m => throw new IllegalStateException(m))
```
Is there a better way of dealing with errors rather than using exceptions?  Could we enforce type safety so this fails at compile time rather than run time?

rtyley replied:
> Is there a better way of dealing with errors rather than using exceptions? Could we enforce type safety so this fails at compile time rather than run time?

I think compile-time safety would be difficult, because the variants are defined with given countrygroups, and then you have to invert the map in order to find out which countries are not covered. Perhaps there is some kind of hardcore scalaz way of doing this, but I don't know what it is! However, it would be easy to create a unit test to ensure a bad setup fails at build time.

pvighi replied:
Yeah I don't like this either.
 Since I'm already adding unit tests I'm adding one that checks that the generated variant weights make sense.

markjamesbutler replied:
👍  - Agreed, I'd definitely be interested in understanding how to implement these checks at compile time this in future.

pvighi committed:
fixed some bugs, refactored and added tests

pvighi committed:
undoing left over changes to app loader

pvighi committed:
Make sure that variant ranges by country are returned in the right order

rtyley committed:
Simplify sum expression

alexduf commented on app/views/support/ABTest.scala:
```diff
@@ -38,7 +38,7 @@ trait TestTrait {
    def addRanges(filteredVariants: Set[Variant]): Seq[(Double, Variant)] = {
      val filteredVariantList = filteredVariants.toList
-     val weightSum: Double = filteredVariantList.map(_.weight).fold(0.0)(_ + _)
+     val weightSum: Double = filteredVariantList.map(_.weight).sum
```
:+1: that looks nicer :)

pvighi committed:
typo in unit test case name

pvighi committed:
Merge branch 'variantsByRegion' of github.com:guardian/contributions-frontend into variantsByRegion

rtyley committed:
Use scanLeft for clearer accumulation of CDF

Also make division by `totalWeight` the final operation, to reduce
rounding errors.

rtyley commented:
:+1:

pvighi commented:
Seems to work fine ( but we have no country specific variants configured yet)
